src/Features/Core/Portable/NavigationBar/AbstractNavigationBarItemService.cs

```diff
@@ -35,15 +39,57 @@ public async Task<ImmutableArray<RoslynNavigationBarItem>> GetItemsAsync(Documen
             return items;
         }
 
-        protected static TextSpan? GetSelectionSpan(ISymbol symbol, SyntaxTree tree)
+        protected static SymbolItemLocation? GetSymbolLocation(
+            Solution solution, ISymbol symbol, SyntaxTree tree, Func<SyntaxReference, TextSpan> computeFullSpan)
         {
-            foreach (var location in symbol.Locations)
+            return GetSymbolLocation(solution, symbol, tree, computeFullSpan, symbol.DeclaringSyntaxReferences);
+        }
+
+        private static SymbolItemLocation? GetSymbolLocation(
+            Solution solution, ISymbol symbol, SyntaxTree tree,
+            Func<SyntaxReference, TextSpan> computeFullSpan,
+            ImmutableArray<SyntaxReference> allReferences)
+        {
+            if (allReferences.Length == 0)
+                return null;
+
+            // See if there are any references in the starting file. We always prefer those for any symbol we find.
+            var referencesInCurrentFile = allReferences.WhereAsArray(r => r.SyntaxTree == tree);
+            if (referencesInCurrentFile.Length > 0)
```

Comment:
is there a possible case where there are caret locations in the current document, but the navigation action should take you to a different document?  

It seems like the answer is no for partial types (navigation just takes you to the declaration in the file).  And for partial methods it looks like currently 2 different items in the navbar, one for the declaration and one for the implementation.  So I'm guessing this case doesn't matter but wanted to confirm

Reply:
> is there a possible case where there are caret locations in the current document, but the navigation action should take you to a different document?

I can technically think of a case.  You could technically have a symbol that had a part in the current file that was totally hidden.  In that case we should instead navigate to the location in another file that was not hidden...

However, we didn't really support that before, so i'm ok with this limitation staying.